kitsune/notifications/models.py

```diff
@@ -1,13 +1,77 @@
 from datetime import datetime
 
 from django.contrib.auth.models import User
+from django.contrib.contenttypes.models import ContentType
 from django.db import models
+from django.db.models import Q
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+import actstream.registry
+from actstream.models import Action, Follow
 
 from kitsune.sumo.models import ModelBase
 
 
-class PushNotificationRegistration(ModelBase):
+class Notification(ModelBase):
+    owner = models.ForeignKey(User, db_index=True)
+    action = models.ForeignKey(Action)
+    read_at = models.DateTimeField(blank=True, null=True)
+
+    @property
+    def is_read(self):
+        return self.read_at is not None
+
+    @is_read.setter
+    def is_read(self, newval):
+        oldval = self.read_at is not None
+        if not oldval and newval:
+            self.read_at = datetime.now()
+        elif oldval and not newval:
+            self.read_at = None
 
+
+class PushNotificationRegistration(ModelBase):
     creator = models.ForeignKey(User, db_index=True)
     created = models.DateTimeField(default=datetime.now)
     push_url = models.CharField(max_length=256)
+
+
+actstream.registry.register(User)
+
+
+@receiver(post_save, sender=Action, dispatch_uid='action_create_notifications')
+def add_action_for_new_question(sender, instance, created, **kwargs):
+    """When an Action is created, notify every user following something in the Action."""
+    if not created:
+        return
+
+    # Make sure that actstream is cool with all these types of objects.
+
+    # For each attribute of the action, check that the attribute is valid, and
+    # build a query that finds all Follow objects that match it.
+    actstream.registry.check(instance.actor)
+    query = Q(
+        content_type=ContentType.objects.get_for_model(instance.actor).pk,
+        object_id=instance.actor.pk)
+
+    if instance.target is not None:
+        actstream.registry.check(instance.target)
+        query = query | Q(
+            content_type=ContentType.objects.get_for_model(instance.target).pk,
+            object_id=instance.target.pk,
+            actor_only=False)
+
+    if instance.action_object is not None:
+        actstream.registry.check(instance.action_object)
+        query = query | Q(
+            content_type=ContentType.objects.get_for_model(instance.action_object).pk,
+            object_id=instance.action_object.pk,
+            actor_only=False)
+
+    # execute the above query, iterate through the results, get every user
+    # assocated with those Follow objects, and fire off a  notification to
+    # every one of them. Use a set to only notify each user once.
+    users_to_notify = set(f.user for f in Follow.objects.filter(query))
+    notifications = [Notification(owner=u, action=instance) for u in users_to_notify]
+    Notification.objects.bulk_create(notifications)
```

Comment:
Until just now I (for some reason) thought that signal handlers were async, and so they wouldn't need celery tasks. I've learned I'm wrong. I'll fix this.